File: utilities_js/crafting.js
/**
 * Crafting Module for Our Legacy (Browser Version)
 * Alchemy and crafting functionality
 * Ported from utilities/crafting.py
 */

import { Colors } from './settings.js';
import { getRarityColor } from './shop.js';

/**
 * Visit the Alchemy workshop
 * @param {Object} game - Game instance
 * @param {Function} askFunc - Ask function for input
 * @returns {Promise<void>}
 */
export async function visitAlchemy(game, askFunc) {
  if (!game.player) {
    game.print(game.lang?.get("no_character") || "No character created yet.");
    return;
  }

  const recipes = game.craftingData?.recipes || {};
  if (!recipes || Object.keys(recipes).length === 0) {
    game.print(game.lang?.get('ui_no_crafting_recipes') || "No crafting recipes available.");
    return;
  }

  game.print(`\n${Colors.MAGENTA}${Colors.BOLD}=== ALCHEMY WORKSHOP ===${Colors.END}`);
  game.print("Welcome to the Alchemy Workshop! Here you can craft potions, elixirs, and items.");
  game.print(`\nYour gold: ${Colors.GOLD}${game.player.gold}${Colors.END}`);

  displayCraftingMaterials(game);

  while (true) {
    game.print("\n=== ALCHEMY WORKSHOP ===");
    game.print("Categories: [P]otions, [E]lixirs, [N]chantments, [U]tility, [A]ll");
    game.print("[C]raft item, [M]aterials, [B]ack");

    const choice = (await askFunc("\nChoose an option: ")).trim().toUpperCase();

    if (choice === 'B' || !choice) {
      break;
    } else if (choice === 'P') {
      await displayRecipesByCategory(game, 'Potions', askFunc);
    } else if (choice === 'E') {
      game.print("Elixirs or Enchantments? [E/N]:");
      const sub = (await askFunc("Choose (E/N): ")).trim().toUpperCase();
      if (sub === 'E') {
        await displayRecipesByCategory(game, 'Elixirs', askFunc);
      } else if (sub === 'N') {
        await displayRecipesByCategory(game, 'Enchantments', askFunc);
      }
    } else if (choice === 'U') {
      await displayRecipesByCategory(game, 'Utility', askFunc);
    } else if (choice === 'A') {
      await displayAllRecipes(game, askFunc);
    } else if (choice === 'C') {
      await craftItem(game, askFunc);
    } else if (choice === 'M') {
      displayCraftingMaterials(game);
    } else {
      game.print("Invalid choice.");
    }
  }
}

export function displayCraftingMaterials(game) {
  if (!game.player) return;

  game.print("\n=== Your Materials ===");
  const materialCategories = game.craftingData?.material_categories || {};
  const allMaterials = new Set();
  for (const materials of Object.values(materialCategories)) {
    for (const mat of materials) allMaterials.add(mat);
  }

  const materialCounts = {};
  for (const item of game.player.inventory) {
    if (allMaterials.has(item)) {
      materialCounts[item] = (materialCounts[item] || 0) + 1;
    }
  }

  if (Object.keys(materialCounts).length === 0) {
    game.print("You have no crafting materials.");
    return;
  }

  game.print(`${'Material'.padEnd(25)} ${'Quantity'.padEnd(10)}`);
  game.print("-".repeat(35));
  for (const [material, count] of Object.entries(materialCounts).sort()) {
    game.print(`${material.padEnd(25)} ${count.toString().padEnd(10)}`);
  }
}

export async function displayRecipesByCategory(game, category, askFunc) {
  const recipes = game.craftingData?.recipes || {};
  const categoryRecipes = Object.entries(recipes).filter(([_, rdata]) => rdata.category === category);

  if (categoryRecipes.length === 0) {
    game.print(`No recipes for ${category}.`);
    return;
  }

  game.print(`\n${Colors.BOLD}=== ${category.toUpperCase()} ===${Colors.END}`);
  for (let i = 0; i < categoryRecipes.length; i++) {
    const [rid, rdata] = categoryRecipes[i];
    const name = rdata.name || rid;
    const rarity = rdata.rarity || 'common';
    const rarityColor = getRarityColor(rarity);
    game.print(`${i + 1}. ${rarityColor}${name}${Colors.END}`);
  }
}

export async function displayAllRecipes(game, askFunc) {
  const recipes = game.craftingData?.recipes || {};
  if (Object.keys(recipes).length === 0) {
    game.print("\nNo recipes available.");
    return;
  }

  const pageSize = 10;
  const recipeList = Object.entries(recipes);
  let currentPage = 0;

  while (true) {
    const start = currentPage * pageSize;
    const end = start + pageSize;
    const pageItems = recipeList.slice(start, end);

    game.print("\n=== All Recipes ===");
    for (let i = 0; i < pageItems.length; i++) {
      const [rid, rdata] = pageItems[i];
      const name = rdata.name || rid;
      const category = rdata.category || 'Unknown';
      const rarityColor = getRarityColor(rdata.rarity || 'common');
      game.print(`${start + i + 1}. ${rarityColor}${name}${Colors.END} (${category})`);
    }

    const totalPages = Math.ceil(recipeList.length / pageSize);
    game.print(`\nPage ${currentPage + 1}/${totalPages}`);
    if (currentPage > 0) game.print("P. Previous Page");
    if (currentPage < totalPages - 1) game.print("N. Next Page");
    game.print("C. Craft Option, B. Back");

    const choice = (await askFunc("\nChoose an option: ")).trim().toUpperCase();
    if (choice === 'B') break;
    else if (choice === 'N' && currentPage < totalPages - 1) currentPage++;
    else if (choice === 'P' && currentPage > 0) currentPage--;
    else if (choice === 'C') await craftItem(game, askFunc);
  }
}

export async function craftItem(game, askFunc) {
  const recipes = game.craftingData?.recipes || {};
  const recipeNames = Object.keys(recipes);

  game.print("\n=== Craft Item ===");
  recipeNames.forEach((rid, i) => {
    const rdata = recipes[rid];
    game.print(`${i + 1}. ${getRarityColor(rdata.rarity || 'common')}${rdata.name || rid}${Colors.END}`);
  });

  const choice = (await askFunc(`\nChoose (1-${recipeNames.length}) or Enter: `)).trim();
  if (!choice || !/^\d+$/.test(choice)) return;

  const idx = parseInt(choice) - 1;
  const recipe = recipes[recipeNames[idx]];
  if (!recipe) return;

  if (game.player.level < (recipe.skill_requirement || 1)) {
    game.print(`\n${Colors.RED}Need level ${recipe.skill_requirement || 1}!${Colors.END}`);
    return;
  }

  const materialsNeeded = recipe.materials || {};
  for (const [material, quantity] of Object.entries(materialsNeeded)) {
    if (game.player.inventory.filter(x => x === material).length < quantity) {
      game.print(`Missing ${material}`);
      return;
    }
  }

  const confirm = (await askFunc("\nCraft? (y/n): ")).trim().toLowerCase();
  if (confirm !== 'y') return;

  for (const [material, quantity] of Object.entries(materialsNeeded)) {
    for (let i = 0; i < quantity; i++) {
      game.player.inventory.splice(game.player.inventory.indexOf(material), 1);
    }
  }

  const outputItems = recipe.output || {};
  for (const [item, quantity] of Object.entries(outputItems)) {
    for (let i = 0; i < quantity; i++) {
      game.player.inventory.push(item);
      game.updateMissionProgress('collect', item);
    }
  }
  game.print(`${Colors.GREEN}Crafted ${recipe.name}!${Colors.END}`);
}

export default {
  visitAlchemy,
  displayCraftingMaterials,
  displayRecipesByCategory,
  displayAllRecipes,
  craftItem,
  visit_alchemy: visitAlchemy,
  display_crafting_materials: displayCraftingMaterials,
  display_recipes_by_category: displayRecipesByCategory,
  display_all_recipes: displayAllRecipes,
  craft_item: craftItem
};

// Also export as named exports for ES6 imports
export { visitAlchemy as visit_alchemy };
export { displayCraftingMaterials as display_crafting_materials };
export { displayRecipesByCategory as display_recipes_by_category };
export { displayAllRecipes as display_all_recipes };
export { craftItem as craft_item };

File: utilities/crafting.py
def visit_alchemy(self):
    from main import Colors, ask, clear_screen
    """Visit the Alchemy workshop to craft items"""
    if not self.player:
        print(self.lang.get("no_character"))
        return

    if not self.crafting_data or not self.crafting_data.get('recipes'):
        print(self.lang.get('ui_no_crafting_recipes'))
        return

    print(
        f"\n{Colors.MAGENTA}{Colors.BOLD}=== ALCHEMY WORKSHOP ==={Colors.END}")
    print(
        "Welcome to the Alchemy Workshop! Here you can craft potions, elixirs, and items."
    )
    print(f"\nYour gold: {Colors.GOLD}{self.player.gold}{Colors.END}")

    # Display available materials from inventory
    self._display_crafting_materials()

    while True:
        clear_screen()
        print(self.lang.get("n_alchemy_workshop"))
        print(
            "Categories: [P]otions, [E]lixirs, [E]ntchantments, [U]tility, [A]ll"
        )
        print(self.lang.get('ui_craft_item'))

        choice = ask("\nChoose an option: ").strip().upper()

        if choice == 'B' or not choice:
            break
        elif choice == 'P':
            self._display_recipes_by_category('Potions')
        elif choice == 'E':
            # Ask which type of E (Elixirs or Enchantments)
            print(self.lang.get('ui_elixirs_enchantments'))
            sub = ask("Choose (E/N): ").strip().upper()
            if sub == 'E':
                self._display_recipes_by_category('Elixirs')
            elif sub == 'N':
                self._display_recipes_by_category('Enchantments')
        elif choice == 'U':
            self._display_recipes_by_category('Utility')
        elif choice == 'A':
            self._display_all_recipes()
        elif choice == 'C':
            self._craft_item()
        elif choice == 'M':
            self._display_crafting_materials()
        else:
            print(self.lang.get("invalid_choice"))


def _display_crafting_materials(self):
    """Display materials available in player's inventory"""
    if not self.player:
        return

    print(self.lang.get("n_your_materials"))

    # Get all material categories
    material_categories = self.crafting_data.get('material_categories', {})

    # Collect all possible materials
    all_materials = set()
    for materials in material_categories.values():
        all_materials.update(materials)

    # Count materials in inventory
    material_counts = {}
    for item in self.player.inventory:
        if item in all_materials:
            material_counts[item] = material_counts.get(item, 0) + 1

    if not material_counts:
        print(self.lang.get('ui_no_crafting_materials'))
        print(
            "Materials can be found as drops from enemies or purchased from shops."
        )
        return

    print(f"{'Material':<25} {'Quantity':<10}")
    print("-" * 35)
    for material, count in sorted(material_counts.items()):
        print(f"{material:<25} {count:<10}")


def _display_recipes_by_category(self, category: str):
    from main import Colors, get_rarity_color
    """Display recipes filtered by category"""
    if not self.crafting_data:
        return

    recipes = self.crafting_data.get('recipes', {})
    category_recipes = [(rid, rdata) for rid, rdata in recipes.items()
                        if rdata.get('category') == category]

    if not category_recipes:
        print(self.lang.get("no_recipes_category").format(category=category))
        return

    print(f"\n{Colors.BOLD}=== {category.upper()} ==={Colors.END}")
    for i, (rid, rdata) in enumerate(category_recipes, 1):
        name = rdata.get('name', rid)
        rarity = rdata.get('rarity', 'common')
        rarity_color = get_rarity_color(rarity)
        print(f"{i}. {rarity_color}{name}{Colors.END}")


def _display_all_recipes(self):
    from main import Colors, ask, get_rarity_color
    """Display all available recipes"""
    if not self.crafting_data:
        return

    recipes = self.crafting_data.get('recipes', {})
    if not recipes:
        print(f"\n{self.lang.get('ui_no_recipes_available')}")
        return

    page_size = 10
    recipe_list = list(recipes.items())
    current_page = 0

    while True:
        start = current_page * page_size
        end = start + page_size
        page_items = recipe_list[start:end]

        print(self.lang.get("n_all_recipes"))
        for i, (rid, rdata) in enumerate(page_items, 1):
            name = rdata.get('name', rid)
            category = rdata.get('category', 'Unknown')
            rarity = rdata.get('rarity', 'common')
            rarity_color = get_rarity_color(rarity)
            print(
                f"{start + i}. {rarity_color}{name}{Colors.END} ({category})")

        total_pages = (len(recipe_list) + page_size - 1) // page_size
        print(f"\nPage {current_page + 1}/{total_pages}")

        if total_pages > 1:
            if current_page > 0:
                print(f"P. {self.lang.get('ui_previous_page')}")
            if current_page < total_pages - 1:
                print(f"N. {self.lang.get('ui_next_page')}")
        print(f"C. {self.lang.get('ui_craft_option')}")
        print(f"B. {self.lang.get('back')}")

        choice = ask("\nChoose an option: ").strip().upper()

        if choice == 'B':
            break
        elif choice == 'N' and current_page < total_pages - 1:
            current_page += 1
        elif choice == 'P' and current_page > 0:
            current_page -= 1
        elif choice == 'C':
            self._craft_item()
        else:
            print(self.lang.get("invalid_choice"))


def _craft_item(self):
    from main import Colors, ask, get_rarity_color
    """Craft an item using materials from inventory"""
    if not self.player or not self.crafting_data:
        print(self.lang.get('ui_cannot_craft'))
        return

    recipes = self.crafting_data.get('recipes', {})

    # Show all recipes for selection
    print(self.lang.get("n_craft_item"))
    recipe_names = list(recipes.keys())

    for i, rid in enumerate(recipe_names, 1):
        rdata = recipes[rid]
        name = rdata.get('name', rid)
        rarity = rdata.get('rarity', 'common')
        rarity_color = get_rarity_color(rarity)
        print(f"{i}. {rarity_color}{name}{Colors.END}")

    choice = ask(
        f"\nChoose recipe (1-{len(recipe_names)}) or press Enter to cancel: "
    ).strip()

    if not choice:
        return

    if not choice.isdigit():
        print(self.lang.get("invalid_choice"))
        return

    idx = int(choice) - 1
    if not (0 <= idx < len(recipe_names)):
        print(self.lang.get('invalid_recipe_number'))
        return

    recipe_id = recipe_names[idx]
    recipe = recipes[recipe_id]

    # Check skill requirement
    skill_req = recipe.get('skill_requirement', 1)
    if self.player.level < skill_req:
        print(
            f"\n{Colors.RED}You need at least level {skill_req} to craft this item.{Colors.END}"
        )
        return

    # Get materials required
    materials_needed = recipe.get('materials', {})

    # Check if player has materials
    missing_materials = []
    for material, quantity in materials_needed.items():
        in_inventory = self.player.inventory.count(material)
        if in_inventory < quantity:
            missing_materials.append(
                f"{material} (need {quantity}, have {in_inventory})")

    if missing_materials:
        print(self.lang.get("nmissing_materials"))
        for m in missing_materials:
            print(f"  - {m}")
        print(f"\n{self.lang.get('ui_gather_more_materials')}")
        return

    # Show craft confirmation
    output_items = recipe.get('output', {})
    print(self.lang.get("n_craft_confirmation"))
    print(self.lang.get("recipe_name_msg").format(name=recipe.get('name')))
    print(
        f"Output: {', '.join(f'{qty}x {item}' for item, qty in output_items.items())}"
    )
    print(f"\n{self.lang.get('ui_materials_consume')}")
    for material, quantity in materials_needed.items():
        print(f"  - {quantity}x {material}")

    confirm = ask("\nCraft this item? (y/n): ").strip().lower()
    if confirm != 'y':
        print(self.lang.get('ui_crafting_cancelled'))
        return

    # Consume materials
    for material, quantity in materials_needed.items():
        for _ in range(quantity):
            self.player.inventory.remove(material)

    # Add crafted items to inventory
    for item, quantity in output_items.items():
        for _ in range(quantity):
            self.player.inventory.append(item)
            self.update_mission_progress('collect', item)

    print(
        f"\n{Colors.GREEN}Successfully crafted {recipe.get('name')}!{Colors.END}"
    )
    for item, quantity in output_items.items():
        print(
            self.lang.get("received_quantity_item").format(quantity=quantity,
                                                           item=item))
